### PR TITLE
8255011: [TESTBUG] compiler/codecache/stress/UnexpectedDeoptimizationAllTest.java timed out

### PR DESCRIPTION
**Pull request by neliasso**
Hi, 

This patch updates the code cache stress tests. I haven't been able to reproduce or retrieve a core file. 

What I can see is that the tests provokes compile storms, and that the single C1 thread (on a 4CPU system) sometimes has trouble keeping up. A factor may also be that the tests run time scale with the timeout time - so that the time allotted as margin before the timeout is only 20% of the total runtime. Combining this with Xcomp, and that the test may run concurrently with other stress tests, it is reasonable that a timeout may occur.

I suggest to cap the tests to 60 seconds of testing. I've experimented with meassuring how much work is done and use that as a metric - but the different tests that use the CodeCacheStressRunner has completely different profiles. 

In UnexpectedDeoptimizationAllTest.java I have adjusted the sleep time to 100 millis between the invalidations of the entire code cache.

In UnexpectedDeoptimizationTest.java  I have added a sleep of 10 miilis between deoptimizing parts of the stack. The idea is to give the stack time to growth a bit before the next deoptimization. Otherwise the test might end up running mostly in the interpreter.

Please review,
Nils Eliasson

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ❌ (1/9 failed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test task**
- [Linux x86 (hs/tier1 compiler)](https://github.com/neliasso/jdk/runs/1357676492)

### Issue
 * [JDK-8255011](https://bugs.openjdk.java.net/browse/JDK-8255011): [TESTBUG] compiler/codecache/stress/UnexpectedDeoptimizationAllTest.java timed out

### Reviewers
 * [Igor Ignatyev](https://openjdk.java.net/census#iignatyev) (@iignatev - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1030/head:pull/1030`
`$ git checkout pull/1030`

### CODE REVIEW

**Comment by neliasso** [comment]
/summary compiler/codecache/stress/UnexpectedDeoptimizationAllTest.java timed out

File: test/hotspot/jtreg/compiler/codecache/stress/CodeCacheStressRunner.java [hunk 1]
```diff
@@ -27,17 +27,21 @@
...
         Helper.startInfiniteLoopThread(action);
         try {
-            // adjust timeout and substract vm init and exit time
+            // Adjust timeout and substract vm init and exit time
+            // Runs with fixed workload that is capped by the TimeLimitedRunner
+            end_time = System.nanoTime() + 60 * 1000 * 1000 * 1000L;
             long timeout = Utils.adjustTimeout(Utils.DEFAULT_TEST_TIMEOUT);
-            timeout *= 0.8;
+            timeout *= 0.75;
             new TimeLimitedRunner(timeout, 2.0d, this::test).call();
```
Comment (iignatev) [inline test/hotspot/jtreg/compiler/codecache/stress/CodeCacheStressRunner.java:45]: why do you need `end_time`? won't it be enough to just set `timeout` to 60_000?

**Review by iignatev — CHANGES_REQUESTED** [review]

**Comment by iignatev** [comment]
@lepestock, IIRC, you are/were working on fixing timeouts in some other code cache tests which use `compiler/codecache/stress/CodeCacheStressRunner`, so I believe this might be of interest to you.

**Comment by iignatev** [comment]
@neliasso ,

could you please explain why in `UnexpectedDeoptimizationAllTest` you "have adjusted the sleep time to 100 millis between the invalidations of the entire code cache"?

**Comment by neliasso** [comment]
/summary [TESTBUG] UnexpectedDeoptimizationAllTest.java timed out

> @neliasso ,
> 
> could you please explain why in `UnexpectedDeoptimizationAllTest` you "have adjusted the sleep time to 100 millis between the invalidations of the entire code cache"?

I was experimenting with different levels of contention. 10 millis + Xcomp gets 30% fewer methods compiled, but in all other cases 10 millis results in more compilations. It's a toss for me.

File: test/hotspot/jtreg/compiler/codecache/stress/CodeCacheStressRunner.java (hunk 1 above)
Comment (neliasso) [reply test/hotspot/jtreg/compiler/codecache/stress/CodeCacheStressRunner.java]: I had a much more complex version at first. I realize that now it is equivalent to just setting 60 secs. 

I have fixed in in the update.

**Comment by lepestock** [comment]
@neliasso, could you please explain how those compile storms cause timeouts? As far as I could find, the JVM doesn't wait for the threads to finish, it gives them approx. [10 seconds](http://hg.openjdk.java.net/jdk/jdk/file/ee1d592a9f53/src/hotspot/share/runtime/vmOperations.cpp#l388), and then just [exits](http://hg.openjdk.java.net/jdk/jdk/file/ee1d592a9f53/src/hotspot/share/runtime/vmOperations.cpp#l424).

So we only have to make sure that our 20% is larger than 10s + some reasonable margin, right?

**Comment by neliasso** [comment]
> @neliasso, could you please explain how those compile storms cause timeouts? As far as I could find, the JVM doesn't wait for the threads to finish, it gives them approx. [10 seconds](http://hg.openjdk.java.net/jdk/jdk/file/ee1d592a9f53/src/hotspot/share/runtime/vmOperations.cpp#l388), and then just [exits](http://hg.openjdk.java.net/jdk/jdk/file/ee1d592a9f53/src/hotspot/share/runtime/vmOperations.cpp#l424).

I don't know exactly since 1) I can't reproduce it 2) I didn't get any core or thread dump from the failure.

What I do know is 1) The failure was on a slow platform 2) It was probably running concurrently with other tests 3) It was running with -Xcomp

Using JFR and jcmd I have seen - 1) very long compile queues (> 500 for C1)

I did some measurements on workload too. In normal mode 6500 iterations of CodeCacheStressRunner was performed in 30 sec, with -Xcomp only 250. But in total compiles -Xcomp had 30% less compilations.

What I also do know is that the CodeCache lock will be contended by the compile threads and the test thread that is invalidating everything in the cache. With long compile queues the MethodCompileQueue lock can also become contented because the time held is proportional to the number of compile tasks in the queue. In register_method both locks are held.

> 
> So we only have to make sure that our 20% is larger than 10s + some reasonable margin, right?
>

Yes. I suggest 60 seconds test time and minimum of 60 seconds margin. 
I also suggest to not scale the test time with the timeout fraction since the run times becomes excessive. I experimented with keeping the CodeCaceStressRunner iterations constant but the different tests have completely different profiles and the variants are too many. 60 seconds is still a huge amount of compiles and invalidations. (times 2 since all tests are run with and without segmented code cache).

**Comment by iignatev** [comment]
> > @neliasso ,
> > could you please explain why in `UnexpectedDeoptimizationAllTest` you "have adjusted the sleep time to 100 millis between the invalidations of the entire code cache"?
> 
> I was experimenting with different levels of contention. 10 millis + Xcomp gets 30% fewer methods compiled, but in all other cases 10 millis results in more compilations. It's a toss for me.
> 
> I reverted since it doesn't really matter.

👍 , but could you please add comments to both `sleep` so the future readers would know why they are there?

**Comment by iignatev** [comment]
>... In normal mode 6500 iterations of CodeCacheStressRunner was performed in 30 sec, with -Xcomp only 250. But in total compiles -Xcomp had 30% less compilations...

would it make sense to update `CodeCacheStressRunner` and/or individual tests to count iterations and have a reasonable (test-specific?) threshold below which a test isn't considered passed, but instead considered skipped (ie throws `jtreg.SkippedException`). this way we would know if tests run enough iterations in the allocated amount of time, and if they don't in some configuration, we will have data to make proper decision on either increasing time budget,  decreasing threshold, or excluding configurations.

-- Igor

**Comment by neliasso** [comment]
> > ... In normal mode 6500 iterations of CodeCacheStressRunner was performed in 30 sec, with -Xcomp only 250. But in total compiles -Xcomp had 30% less compilations...
> 
> would it make sense to update `CodeCacheStressRunner` and/or individual tests to count iterations and have a reasonable (test-specific?) threshold below which a test isn't considered passed, but instead considered skipped (ie throws `jtreg.SkippedException`). this way we would know if tests run enough iterations in the allocated amount of time, and if they don't in some configuration, we will have data to make proper decision on either increasing time budget, decreasing threshold, or excluding configurations.
> 
> -- Igor

I experimented with couting the CodeCaceStressRunner iterations but the different tests have completely different profiles and the variants are many - 4 tests, 2 modes (Xcomp) that all need separate values.

I would like to see that we added thread dumps and core files on timeouts. Then we can experiment with lower timeouts since it will be possible to diagnose them quickly.

**Comment by iignatev** [comment]
>  would like to see that we added thread dumps and core files on timeouts.

we actually do that since JDK 9, [JEP 279](https://bugs.openjdk.java.net/browse/JDK-8075621) added the failure handler which runs many things including `jstack`, `gdb -ex thread apply all backtrace` (or its equivalent) to get java and native thread dumps and `kill -ABRT` to generate core files on timeouts. b/c it's run concurrently to a timed out process, you don't always get the data. so I'm wondering what you meant.

-- Igor

**Comment by neliasso** [comment]
> we actually do that since JDK 9, [JEP 279](https://bugs.openjdk.java.net/browse/JDK-8075621) added the failure handler which runs many things including `jstack`, `gdb -ex thread apply all backtrace` (or its equivalent) to get java and native thread dumps and `kill -ABRT` to generate core files on timeouts. however b/c it's run concurrently to a timed out process, you don't always get the data. so I'm wondering what you meant.

That has completely passed me by. I must look into it.

Regarding the test.

I think 60 secs is plenty of time for this test, especially since it runs multiple times. If there is a deadlock - the test will timeout regardless of timeout time. Adding some workload measurement would be nice - but not worth the time in my opinion.

**Comment by iignatev** [comment]
> Adding some workload measurement would be nice - but not worth the time in my opinion.

sure-sure, I didn't mean to do it as part of this PR, just a thought for future improvements in these tests.

**Review by iignatev — APPROVED** [review]
LGTM. however I'd like to you to wait for explcit 'ok' from Evgeny (@lepestock) as well.

BTW, you need to either update the title of this PR to `[TESTBUG] compiler/codecache/stress/UnexpectedDeoptimizationAllTest.java timed out` or change [8255011](https://bugs.openjdk.java.net/browse/JDK-8255011)'s title to `[TESTBUG] UnexpectedDeoptimizationAllTest.java timed out`

-- Igor

**Comment by lepestock** [comment]
LGTM. I was considering similar change as well, so I agree.

**Comment by neliasso** [comment]
/summary Change CodeCacheStressRunner to have a 60 second test time

**Comment by neliasso** [comment]
Thank you Evgeny and Igor!

**Pull request closed without merging**